lib/hex/hex.cpp

```diff
@@ -0,0 +1,119 @@
+// EVMC: Ethereum Client-VM Connector API.
+// Copyright 2021 The EVMC Authors.
+// Licensed under the Apache License, Version 2.0.
+
+#include <evmc/hex.hpp>
+#include <cctype>
+
+namespace evmc
+{
+namespace
+{
+inline int from_hex_digit(char h)
+{
+    if (h >= '0' && h <= '9')
+        return h - '0';
+    else if (h >= 'a' && h <= 'f')
+        return h - 'a' + 10;
+    else if (h >= 'A' && h <= 'F')
+        return h - 'A' + 10;
+    else
+        throw hex_error{hex_errc::invalid_hex_digit};
+}
+
+template <typename OutputIt>
+inline void from_hex(const std::string& hex, OutputIt result)
+{
+    // TODO: This can be implemented with hex_decode_iterator and std::copy.
+
+    // Omit the optional 0x prefix.
+    const auto hex_begin =
+        (hex.size() >= 2 && hex[0] == '0' && hex[1] == 'x') ? hex.begin() + 2 : hex.begin();
+
+    constexpr int empty_byte_mark = -1;
+    int b = empty_byte_mark;
+    for (auto it = hex_begin; it != hex.end(); ++it)
+    {
+        const auto h = *it;
+        if (std::isspace(h))
```

Comment:
Hm why do you want to allow whitespace? If this is to skip extra whitespace from command line, it's better to handle it in command line code.

Also currently it looks to allow `"0x12"` and `"  12"` (also `"1 2"`), but not `"   0x12"`

Reply:
Yes, this is what I want for simplicity. This is useful for loading hex from files to get rid of line breaks.

Reply:
Just `hex.erase(std::remove_if(hex.begin(), hex.end()), std::isspace), hex.end())` them!

Reply:
This will not work with input iterators (planned change). In the end we can have `hex_decoding_iterator` and `strip_whitespace_iterator` but I think having just one is simpler.